vms/proposervm/proposer/windower.go

```diff
@@ -148,3 +165,108 @@ func (w *windower) Delay(ctx context.Context, chainHeight, pChainHeight uint64,
 	}
 	return delay, nil
 }
+
+func (w *windower) ExpectedProposer(
+	ctx context.Context,
+	blockHeight,
+	pChainHeight uint64,
+	parentBlockTime,
+	blockTime time.Time,
+) (ids.NodeID, error) {
+	validators, err := w.sortedValidators(ctx, pChainHeight)
+	if err != nil {
+		return ids.EmptyNodeID, err
+	}
+
+	var (
+		slot             = uint64(blockTime.Sub(parentBlockTime) / WindowDuration)
+		seed             = blockHeight ^ bits.Reverse64(slot) ^ w.chainSource
+		source           = prng.NewMT19937_64()
+		validatorWeights = validatorsToWeight(validators)
+	)
+
+	sampler := sampler.NewDeterministicWeightedWithoutReplacement(source)
+	if err := sampler.Initialize(validatorWeights); err != nil {
+		return ids.EmptyNodeID, err
+	}
+
+	source.Seed(seed)
+	indices, err := sampler.Sample(1)
+	if err != nil {
+		return ids.EmptyNodeID, fmt.Errorf("%w, %w", err, ErrNoProposersAvailable)
+	}
+	return validators[indices[0]].id, nil
+}
+
+func (w *windower) MinDelayForProposer(
+	ctx context.Context,
+	chainHeight,
+	pChainHeight uint64,
+	parentBlockTime time.Time,
+	nodeID ids.NodeID,
+	startTime time.Time,
+) (time.Duration, error) {
+	validators, err := w.sortedValidators(ctx, pChainHeight)
+	if err != nil {
+		return 0, err
+	}
+
+	var (
+		slot = uint64(startTime.Sub(parentBlockTime) / WindowDuration)
+
+		source           = prng.NewMT19937_64()
+		validatorWeights = validatorsToWeight(validators)
+		sampler          = sampler.NewDeterministicWeightedWithoutReplacement(source)
+	)
+	if err := sampler.Initialize(validatorWeights); err != nil {
+		return 0, err
+	}
+
+	for time.Duration(slot)*WindowDuration < MaxLookAheadWindow {
```

Comment:
Do we have a benchmark on how long this full iteration takes for the default `MaxLookAheadWindow`?

Reply:
(I'm assuming negligible but would be nice to confirm)

Reply:
The time (outside of fetching the validator set) is around 1 ms:
```
goos: darwin
goarch: arm64
pkg: github.com/ava-labs/avalanchego/vms/proposervm/proposer
BenchmarkMinDelayForProposer-12    	    1310	    905543 ns/op	   10693 B/op	     745 allocs/op
PASS
coverage: 9.3% of statements in all
ok  	github.com/ava-labs/avalanchego/vms/proposervm/proposer	2.980s
```
I think this is fine - if we wanted to reduce the maximum number of slots to iterate we could cut this back... But I think we'll fine that the majority of the cost is fetching the validator set fairly quickly.